hips/hip-0016.md

```diff
@@ -0,0 +1,110 @@
+
+With the proposed `export-values` directive, users are able to specify in the **parent** chart (snippet below) which and where values should be exported to the **subchart**.
+
+```yaml
+# Chart.yaml:
+dependencies:
+  - name: client
+    version: 1.0.0
+    export-values:
+    - parent: "port"
+      child: "serverPort"
+  - name: server
+    version: 1.0.0
+    export-values:
+    - parent: "port"
+      child: "exposePort"
+    - "server-config"
+```
+
+```yaml
+# values.yaml
+port: 8080
+exports:
+  server-config:
+    debug: true
+```
+
+```yaml
+# charts/client/values.yaml
+serverPort: 9090
+```
+
+```yaml
+# charts/server/values.yaml
+exposePort: 80
+debug: false
+```
+
+### Value Precedence
+
+Because later changing the order of precedence between value-sources would be a breaking change (same input, but potentially different output), we should agree on an intuitive and user-friendly order of precedence before implementing this HIP.
+
+Suggested precedence:
+
+1. Explicitly set for subchart value (`client.port`)
+1. Explicitly set for parent chart's value (`port`)
+1. default for parent chart's value (`values.yaml` in parent chart)
+1. default for subchart value (`values.yaml` in subchart)
+
+For the example above, this would mean that `.Values.port`, `.Values.client.serverPort` and `.Values.server.exposePort` will default to `8080` unless e.g. `--set port=1234` is provided via flags, which would overwrite all these values.
+In contrast using e.g. `--set client.serverPort=42` would only overwrite the `.Values.client.serverPort` value.
+
+### Combination with `import-values`
+
+It should be able to use both, `export-values` and `import-values` independently and in combination. It should also be possible to export a value which was imported from a third chart.
+
+## Motivation
+
+Currently, it is not possible to compose charts using dependencies without leaking the structure of the values of the subcharts used.
+Users of a parent chart instead have to provide values to the subcharts directly. Alternatively, users can provide `global` values, but in that case the subcharts need to be aware of the fact that they are used as a dependency. In addition, `global` values are available globally as the name suggests and hence multiple subcharts have to have aligned value keys. As a result, the parent chart and its dependencies are rather tightly coupled.
+
+This HIP would allow to freely structure the values the user interacts with.
+
+## Rationale
+
+Using the same parent chart as in the [specification](#specification) without `export-value`, the `port` value must match and must be specified twice by the user.
+
+```yaml
+# values.yaml
+client:
+  serverPort: 8080
+
+server:
+  exposePort: 8080
+```
+
+The proposed structure is not only cleaner for the user, but also less error prone since the port value is only maintained once.
+
+## Backwards Compatibility
+
+This change won't break any existing Helm chart since it only adds new mechanisms.
```

Comment:
How will older Helm clients (e.g. v3.6.2) work with charts generated with these properties. Older versions of Helm are in wide use.

Reply:
This is indeed a tricky problem.
My first idea was: We need to bump the `apiVersion` of the `Chart.yaml` and make the feature only available, if the newer version is used. But I noticed that my local helm installation happily templates my dummy chart, even if I set `apiVersion: v100` in the `Chart.yaml` (which IMHO is a bug).

One alternative I see is recommending to chart authors who use the feature (in the docs of that feature) to include something like this in their templates (maybe in `min-version.yaml` by convention):

```yaml
{{- if semverCompare ">3.10.0" .Capabilities.HelmVersion.Version }}
{{- fail "This chart requires helm version 3.10.0 or higher" }}
{{- end }}
```

`3.10.0` is obviously only a placeholder for whatever version we will include this in.

Another way could be to make it only available, if you target a Kubernetes version to high for older helm versions (as defined in [Supported Version Skew](https://helm.sh/docs/topics/version_skew/#supported-version-skew)) by setting `kubeVersion` in the `Chart.yaml`. But this would be rather hard to document and I also don't think that the helm cli enforces this in any way. But at least it is documented as "not recommended" for a long time.

Reply:
See https://github.com/helm/community/pull/242#pullrequestreview-1026427343